File: bin/Data/EditorLua/view/MenuBar.lua
local EditorUtils = require("EditorLua/core/EditorUtils")
local Translator = require("EditorLua/core/Translate")
local MenuBar = {}
MenuBar.menuTree = {children = {}}
MenuBar.app = nil

function MenuBar.onOpen()
    print("onOpenFile")
end
function MenuBar.onSave()
    --print("onOpen")
end

function MenuBar.onUndo()
    print("onUndo")
end
function MenuBar.onRedo()
    print("onRedo")
end

function MenuBar:init()
    self:addMenu("file/open", self.onOpen)
    self:addMenu("file/save", self.onSave)
    self:addMenu("edit/undo", self.onUndo)
    self:addMenu("edit/redo", self.onRedo)
    self:addMenu("view/ToolBar", self.app.ShowToolBar)
    self:addMenu("view/NodeTree", self.app.ShowNodeTree)
    self:addMenu("view/ResTree", self.app.ShowResTree)
    self:addMenu("view/ResPreview", self.app.ShowResPreview)
    self:addMenu("view/SceneView", self.app.ShowSceneView)
    self:addMenu("view/Inspector", self.app.ShowInspector)
    print("init MenuBar")
end

function MenuBar:addMenu(path,func)
    local pathItems = EditorUtils.StrSplit(path, "/")
    if #pathItems < 2 then
        return
    end
    local curTreeNode = self.menuTree
    for index = 1,#pathItems -1,1 do
        local item = pathItems[index]
        if curTreeNode.children[item] then
            curTreeNode = curTreeNode.children[item]
        else
            curTreeNode.children[item] = {}
            curTreeNode.children[item].children = {}
            curTreeNode.children[item].name = item
            curTreeNode = curTreeNode.children[item]
        end
    end
    --leaf node
    local leafName = pathItems[#pathItems]
    curTreeNode.children[leafName] = {}
    curTreeNode.children[leafName].children = {}
    curTreeNode.children[leafName].name = leafName
    curTreeNode.children[leafName].func = func
end

local renderMenu
renderMenu = function(menuNode)
    for key,value in pairs(menuNode.children) do
        local localName = Translator:translate(value.name)
        if value.func then
            local selected = false
            if imgui.MenuItem(localName,nil,selected) then
                value.func()
            end
        else
            if imgui.BeginMenu(localName) then
                renderMenu(value)
                imgui.EndMenu()
            end
        end
    end
end

function MenuBar:update()
    if imgui.BeginMenuBar() then
        renderMenu(self.menuTree)
        imgui.EndMenuBar()
    end
end

return MenuBar

File: Source/UrhoEditor/views/Menubar.cpp
#include "Menubar.h"
#include "imgui.h"

namespace Urho3DEditor 
{

Menubar::Menubar():_menuTree(new MenuNode())
{
	Init();
}
Menubar::~Menubar() 
{
}

void Menubar::Init() 
{
	AddMenu("file/open",&OnOpen);
	AddMenu("file/save",&OnSave);
	AddMenu("edit/undo",&OnUndo);
	AddMenu("edit/redo",&OnRedo);
	AddMenu("view/ToolBar",&ShowToolBar);
	AddMenu("view/NodeTree",&ShowNodeTree);
	AddMenu("view/ResTree",&ShowResTree);
	AddMenu("view/ResPreview",&ShowResPreview);
	AddMenu("view/SceneView",&ShowSceneView);
	AddMenu("view/Inspector",ShowInspector);
}

void Menubar::AddMenu(const String& path, MenuHandle handle)
{
	auto pathItems = path.Split('/',true);
	if (pathItems.Size()<2)
		return;
	auto curentNode = this->_menuTree;
	for(int index=0; index <pathItems.Size() -1; ++index)
	{
		auto item = pathItems[index];
		if(curentNode->children.Contains(item))
		{
			curentNode = curentNode->children[item];
		}
		else
		{
			curentNode->children[item] = std::shared_ptr<MenuNode>(new MenuNode());
			curentNode->children[item]->name = item;
			curentNode = curentNode->children[item];
		}
	}
	//pathItems中的最后一项是叶子节点
	auto leafName = pathItems[pathItems.Size() - 1];
	curentNode->children[leafName] = std::shared_ptr<MenuNode>(new MenuNode());
	curentNode->children[leafName]->name = leafName;
	curentNode->children[leafName]->_handle = handle;
}

void Menubar::RenderMenu(MenuNode* menuNode)
{
	if(menuNode == nullptr)
		return;
	for(auto item:menuNode->children)
	{
		String localName = item.second_->name;
		if(item.second_->_handle)
		{
			bool selected = false;
			if(ImGui::MenuItem(localName.CString(),"",&selected))
			{
				item.second_->_handle();
			}
		}
		else
		{
			if(ImGui::BeginMenu(localName.CString()))
			{
				RenderMenu(item.second_.get());
				ImGui::EndMenu();
			}
		}
	}
}

void Menubar::Update() 
{
	if (ImGui::BeginMenuBar())
	{
		RenderMenu(_menuTree.get());
		ImGui::EndMenuBar();
	}
}

void Menubar::OnOpen() 
{
	int test = 10;
}

void Menubar::OnSave() {
}

void Menubar::OnUndo() {
}

void Menubar::OnRedo() {
}

void Menubar::ShowToolBar() {
}

void Menubar::ShowNodeTree() {
}

void Menubar::ShowResTree() {
}

void Menubar::ShowResPreview() {
}

void Menubar::ShowSceneView() {
}

void Menubar::ShowInspector() {
}

}